src/libraries/System.Reflection.Emit/src/System/Reflection/Emit/ModuleBuilderImpl.cs

```diff
@@ -180,30 +183,59 @@ internal void AppendMetadata(MethodBodyStreamEncoder methodBodyEncoder)
             }
         }
 
-        private void WriteMethods(TypeBuilderImpl typeBuilder, List<GenericTypeParameterBuilderImpl> genericParams, MethodBodyStreamEncoder methodBodyEncoder)
+        // Need to pre populate all type members tokens so that they can be referenced from other type's metadata
+        private void PrePopulateTypeMembersTokens()
+        {
+            foreach (TypeBuilderImpl typeBuilder in _typeDefinitions)
+            {
+                typeBuilder._firsMethodToken = _nextMethodDefRowId;
+                typeBuilder._firstFieldToken = _nextFieldDefRowId;
+                PrePopulateMethodDefinitionHandles(typeBuilder._methodDefinitions);
+                PrePopulateFieldDefinitionHandles(typeBuilder._fieldDefinitions);
+            }
+        }
+
+        private void PrePopulateFieldDefinitionHandles(List<FieldBuilderImpl> fieldDefinitions)
         {
-            foreach (MethodBuilderImpl method in typeBuilder._methodDefinitions)
+            foreach (FieldBuilderImpl field in fieldDefinitions)
+            {
+                field._handle = MetadataTokens.FieldDefinitionHandle(_nextFieldDefRowId++);
+            }
+        }
+
+        private void PrePopulateMethodDefinitionHandles(List<MethodBuilderImpl> methods)
+        {
+            foreach (MethodBuilderImpl method in methods)
+            {
+                method._handle = MetadataTokens.MethodDefinitionHandle(_nextMethodDefRowId++);
+            }
+        }
+
+        private void WriteMethods(List<MethodBuilderImpl> methods, List<GenericTypeParameterBuilderImpl> genericParams, MethodBodyStreamEncoder methodBodyEncoder)
+        {
+            foreach (MethodBuilderImpl method in methods)
             {
                 int offset = -1;
                 ILGeneratorImpl? il = method.ILGeneratorImpl;
                 if (il != null)
                 {
+                    FillMemberReferences(il);
                     StandaloneSignatureHandle signature = il.Locals.Count == 0 ? default :
                         _metadataBuilder.AddStandaloneSignature(_metadataBuilder.GetOrAddBlob(MetadataSignatureHelper.LocalSignatureEncoder(il.Locals, this)));
                     offset = AddMethodBody(method, il, signature, methodBodyEncoder);
                 }
 
-                MethodDefinitionHandle methodHandle = AddMethodDefinition(method, method.GetMethodSignatureBlob(), offset, _nextParameterRowId);
-                WriteCustomAttributes(method._customAttributes, methodHandle);
-                _nextMethodDefRowId++;
+                MethodDefinitionHandle handle = AddMethodDefinition(method, method.GetMethodSignatureBlob(), offset, _nextParameterRowId);
+                Debug.Assert(method._handle == handle);
```

Comment:
Why was this Assert added? Can it be added to AddMethodDefinition() instead?

Reply:
> Why was this Assert added? 

To make sure the method handle populated with AddMethodDefinition is in sync with prepopulated one.

> Can it be added to AddMethodDefinition() instead?

It could, but `AddMethodDefinition` is a one liner that just calls `MetadataBuilder.AddMethodDefinition`, if you are OK with, don't really want to add another row.